### PR TITLE
etcd: don't confuse prefixes during migration

### LINKED ISSUE #4312: etcd prefix data migration

### PR DESCRIPTION
**Pull request by awly**
The prefix fetching logic has a bug: it treats everything starting with
`/teleport` as the legacy prefix data, even if it's `/teleport-foo/bar`.
This is an issue if user specifies `/teleport-foo` as their custom
prefix. Each restart will copy the data from `/teleport-foo/...` to
`/teleport-foo-foo/...`.

Set the legacy prefix const to `/teleport/` instead. This avoids
excessive copying during startup.

Fixes https://github.com/gravitational/teleport/issues/4312

### CODE REVIEW

**Comment by russjones** [comment]
@awly Can we make a backup of everything under `/teleport/` before we kick off the migration?

**Comment by awly** [comment]
@russjones we can probably backup the data we delete, under the custom prefix (not `/teleport/`).
I'll add that.

**Comment by awly** [comment]
@russjones PTAL

**Review by fspmarshall — APPROVED** [review]
I don't like the idea of having the logic be sensitive to the differences in proceeding/trailing slashes.  Seems like an unnecessary footgun.  Ideally `prefix`, `/prefix`, and `/prefix/` would all mean the same thing from the user's perspective.

Otherwise, LGTM.

**Comment by awly** [comment]
@fspmarshall yeah, I agree.
We made `/prefix` and `prefix` the same, but not with trailing slashes. Now some users have data under `/prefix//foo` so we'll need to be careful when fixing this. That might need another migration :(

**Review by benarent — APPROVED** [review]

**Comment by benarent** [comment]
When we update the docs, I'll come back and update a setup guide for etcd, this does seem to follow the slightly confusing prefix logic of etcd as Forrest has pointed out https://etcd.io/docs/v3.3.12/op-guide/authentication/

btw, we should update https://github.com/gravitational/teleport/issues/2883 to make it clear that it needs the trailing slash incase anyone else finds that ticket. ( or is this only an issue during a migration )

**Comment by fspmarshall** [comment]
> We made `/prefix` and `prefix` the same, but not with trailing slashes. Now some users have data under `/prefix//foo` so we'll need to be careful when fixing this. That might need another migration :(

Huh.  That *is* tricky. Its too bad we don't have a more well-defined format for the keys used by the `Backend` interface.  It would be nice to be able to unambiguously separate path sections.

I suppose that while `/prefix//foo` looks a bit silly, at least it is unambiguous :shrug:

**Comment by awly** [comment]
@benarent actually, all our backed keys start with a `/` so the trailing slash in prefix is not needed.
Updated the description.

Still need approval from @webvictim or @russjones for `/examples/`

**Review by webvictim — APPROVED** [review]

File: lib/backend/etcdbk/etcd.go [hunk 1]
```diff
@@ -687,19 +692,40 @@ func (b *EtcdBackend) syncLegacyPrefix(ctx context.Context) error {
...
+	// First, let's backup the data under configured prefix, in case the
+	// migration kicked in by mistake.
+	backupPrefix := b.backupPrefix(b.cfg.Key)
+	b.Debugf("Backup everything under %q to %q", b.cfg.Key, backupPrefix)
+	for _, kv := range prefixData.Kvs {
+		// Replace the prefix.
+		key := backupPrefix + strings.TrimPrefix(string(kv.Key), b.cfg.Key)
+		b.Debugf("Copying %q -> %q", kv.Key, key)
+		if _, err := b.client.Put(ctx, key, string(kv.Value)); err != nil {
+			errs = append(errs, trace.WrapWithMessage(err, "failed copying %q to %q: %v", kv.Key, key, err))
```
Comment (russjones) [inline lib/backend/etcdbk/etcd.go:709]: Shouldn't we stop migration if this occurs?

Comment (awly) [reply lib/backend/etcdbk/etcd.go:709]: should we crash the teleport process then, or let it proceed with stale/missing data in custom prefix?

Comment (awly) [reply lib/backend/etcdbk/etcd.go]: crashing makes sense, it means we can't write to etcd.
updated.

**Pull request merged**